### ISSUE TITLE
Simple, but powerful, configuration specification

### ISSUE THREAD

**Issue by thorwhalen** [i2mint/py2store#15]
The objective here is to enable a simple but powerful way to manage

- defaults
- configuration specification
- out-of-view secrets (passwords, etc.)

All functions (object constructors and otherwise) have defaults where ever it makes sense to. These defaults are within the code itself. We should strive for flexibility here too though: Allowing the user to modify these defaults to adapt to what's most common to them.

Another use case is that of using and re-using different configurations: We'd like a way for users to be able to cook up configurations that are specific to a given context, name them, and be able to use them later (through its name). 
 
The proposal (to be critiqued) is to use json/yaml files as the specification format. We can use dot-path keys to specify what function they're relevant to (for example `pystore.persister.LocalFilePersister.__init__' to indicate the defaults for the LocalFilePersister() function). Configurations that span many functions can be factored out and applied more broadly. (See https://github.com/thorwhalen/py2api/blob/master/py2api/py2rest/input_trans.py for an example of a json-dialect to implement a "factored" map)

There could be a file in the repository, that specifies the defaults we use there. Further, there could be a mechanism were the user could specify another defaults file that would take precedence over the repositories defaults (still looking into the repository's defaults.json as a fallback for missing defaults). Essentially, chaining these specifications together, one can manage both defaults and named configurations. 

Out-of-view secrets can also be placed in personal json/yaml files since they are part of the arguments that some functions need to build persisters etc.

**Comment by thorwhalen** [comment]
It seems like there should be some URI involved here. Don't know if URIs will bring us the whole way, but at least we should use them as much as we can, since they're known and usually standardized. 

We might have to sometimes represent the URIs by operable python objects though (will all its information parsed and organized), so as to be able to carry out different operations (chaining, merging, etc. ) more readily.

**Comment by Kulv3r** [comment]
An alternative to JSON / YAML config files is a native Python class-based configs like it's usually done in webservices:
```
class BaseConfig:
    LOG_LEVEL = 'INFO'

class LocalFileConfig(BaseConfig):
    PATH = '/somewhere/out/there'
    MODE = 'wb'

class SQLConfig(BaseConfig):
   DB_URI = '...'
   CONNECTION_TIMEOUT = datetime.datetime(seconds=10)

class SQLwSQLAlchemyConfig(SQLConfig):
    SMTH_SPECIFIC = 123
```
The only significant benefit of this approach agains JSON files is ability to use complex objects as config values, like `datetime()` for some timeouts, which is nice, but for sure not a killer feature.

**Comment by thorwhalen** [comment]
@Kulv3r agreed on the basis that it's easier to write python for python (the type problem you talked about, but even simple things like code-completion, etc.

But its advantage is also its disadvantage: i.e, it's python.

The advantage of json/yaml is that it's "language agnostic" so can more readily be written, read, combined, etc. By python or other languages. 

That said, there's often an advantage of transforming the json/yaml to something "more" than a dict. That is, an object specifically meant to contain configs/settings, and operate on them. 

So perhaps we should **start** by building up the "config language" in python, and then once stable, write converters to and from json/yaml (and might even consider supporting other formats (.ini, etc.)

**Comment by thorwhalen** [comment]
https://hackernoon.com/4-ways-to-manage-the-configuration-in-python-4623049e841b

**Comment by thorwhalen** [comment]
py2store is now a legacy forwarding layer over `dol` (its core moved there; see README: "It's advised to use `dol` ... directly when sufficient") and isn't receiving new feature development. The config-specification design (defaults / named configs / secrets) discussed here was never implemented in py2store; `dynaconf`/`python-decouple`/TOML were the leads identified in-thread. Closing -- if this is still wanted, it belongs in `dol` (or a dedicated `dol`-family package), not py2store; not re-filing there per current triage policy, so it'll need to be picked up from here if/when it matters again.
